src/TokenHandler/AccountManager.mo

```diff
@@ -0,0 +1,408 @@
+    public func backlogFunds() : Nat = queuedFunds + underwayFunds;
+
+    /// Notifies of a deposit and schedules backlog processing.
+    /// Returns the newly detected deposit if successful.
+    public func notify(p : Principal) : async* ?Nat {
+      if (isFrozen() or not depositRegistry.lock(p)) return null;
+
+      let latestDeposit = try {
+        await* loadDeposit(p);
+      } catch (err) {
+        depositRegistry.unlock(p);
+        throw err;
+      };
+
+      depositRegistry.unlock(p);
+
+      let prevDeposit = updateDeposit(p, latestDeposit);
+
+      if (latestDeposit < prevDeposit) freezeCallback("latestDeposit < prevDeposit on notify");
+
+      if (latestDeposit > fee_) {
+        // precredit deposit funds
+        ignore creditRegistry.credit(p, latestDeposit - fee_);
+        // schedule consolidation for this p
+        pushToBacklog(p, latestDeposit);
+        // schedule a canister self-call to process the backlog
+        // we need try-catch so that we don't trap if scheduling fails synchronously
+        try ignore processBacklog() catch (_) {};
+      };
+
+      let depositDelta = latestDeposit - prevDeposit : Nat;
+
+      if (depositDelta > 0) journal.push((Time.now(), p, #newDeposit(depositDelta)));
+
+      return ?depositDelta;
+    };
+
+    /// Processes the consolidation transfer for a principal.
+    func processConsolidationTransfer(p : Principal, deposit : Nat) : async* {
+      #Ok : Nat;
+      #Err : ICRC1.TransferError or {
+        #CallIcrc1LedgerError;
+        #InsufficientDeposit;
+      };
+    } {
+      if (deposit <= fee_) return #Err(#InsufficientDeposit);
+
+      let transferAmount : Nat = Int.abs(deposit - fee_);
+
+      let transferResult = try {
+        await icrc1Ledger.icrc1_transfer({
+          from_subaccount = ?Mapping.toSubaccount(p);
+          to = { owner = ownPrincipal; subaccount = null };
+          amount = transferAmount;
+          fee = ?fee_;
+          memo = null;
+          created_at_time = null;
+        });
+      } catch (_) {
+        #Err(#CallIcrc1LedgerError);
+      };
+
+      switch (transferResult) {
+        case (#Ok _) {
+          ignore updateDeposit(p, 0);
+          totalConsolidated_ += transferAmount;
+          journal.push((Time.now(), p, #consolidated({ deducted = deposit; credited = transferAmount })));
+        };
+        case (#Err err) {
+          journal.push((Time.now(), p, #consolidationError(err)));
+        };
+      };
+
+      transferResult;
+    };
+
+    /// Attempts to consolidate the funds for a particular principal.
+    func consolidate(p : Principal) : async* () {
+      let latestDeposit = try { await* loadDeposit(p) } catch (err) {
+        pushToBacklog(p, 0);
+        return;
+      };
+
+      ignore updateDeposit(p, latestDeposit);
+
+      ignore depositRegistry.set(
+        p,
+        func(info) {
+          info.underway := latestDeposit;
+          underwayFunds += latestDeposit - info.underway;
```

Comment:
Should these lines be swapped? In the second line the right hand side will be 0.